### PR TITLE
Env reg variant

### PR DESCRIPTION
**Pull request by memcpy-rand-rand-rand**
### Prerequisites

- [X] I have read and understood the `CONTRIBUTING` guide
- [X] The commit message follows the [conventional commits][cc] guidelines
- [X] Tests for the changes have been added (for bug fixes/features)
- [] Docs have been added / updated (for bug fixes/features)

### Description

refactor(winreg): change getWindowsRegistryKeyValue fn to be more generic, simplify winreg path

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

### CODE REVIEW

**Comment by memcpy-rand-rand-rand** [comment]
Draft PR for discussion.  Added variant-style code to the getWindowsRegistryKeyValue function to make it more useful should others with to use it.

File: src/environment.go [hunk 1]
```diff
@@ -56,6 +56,21 @@ type cache interface {
 	set(key, value string, ttl int)
 }
 
+type windowsRegistryValueType int
+
+const (
```
Comment (JanDeDobbeleer) [inline src/environment.go:61]: should we move these to `environment_windows` so they don't get added for other platforms?

File: src/environment_windows.go [hunk 2]
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
...
+	//
+	// 1. Root key - extract the root HKEY string and turn this into a handle to get started
+	// 2. Path - open this path
+	// 3. Key - get this key value
+	//
+	// If 3 is "" (i.e. the path ends with "\"), then get (Default) key.
+	//
+	regPathParts := strings.SplitN(path, "\\", 2)
 
 	regRootHKeyHandle := getHKEYHandleFromAbbrString(regPathParts[0])
```
Comment (JanDeDobbeleer) [inline src/environment_windows.go:140]: I didn't catch this at first, it assumes there's an element in the array.

File: src/environment_windows.go [hunk 3]
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
...
 	if regRootHKeyHandle == 0 {
 		errorLogMsg := fmt.Sprintf("Error, Supplied root HKEY value not valid: '%s'", regPathParts[0])
 		env.log(Error, "getWindowsRegistryKeyValue", errorLogMsg)
-		return "", errors.New(errorLogMsg)
+		return nil, errors.New(errorLogMsg)
 	}
 
+	// Strip key off the end.
+	lastSlash := strings.LastIndex(regPathParts[1], "\\")
```
Comment (JanDeDobbeleer) [inline src/environment_windows.go:148]: same here, we expect as second element, I would validate if we have all entries before continuing.

File: src/environment_windows.go [hunk 4]
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
...
+
+	regKey := regPathParts[1][lastSlash+1:]
+	regPath := regPathParts[1][0:lastSlash]
+
+	regKeyLogged := regKey
+	if regKeyLogged == "" {
```
Comment (JanDeDobbeleer) [inline src/environment_windows.go]: purely style, I always use `len(regKeyLogged) == 0`

File: src/environment_windows.go [hunk 5]
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
...
+		regKeyLogged = "(Default)"
+	}
+	env.log(Error, "getWindowsRegistryKeyValue", fmt.Sprintf("getWindowsRegistryKeyValue: root:\"%s\", path:\"%s\", key:\"%s\"", regPathParts[0], regPath, regKeyLogged))
+
 	// Second part of split is registry path after HK part - needs to be UTF16 to pass to the windows. API
-	regPathUTF16, regPathUTF16ConversionErr := windows.UTF16FromString(regPathParts[1])
+	regPathUTF16, regPathUTF16ConversionErr := windows.UTF16FromString(regPath)
```
Comment (JanDeDobbeleer) [inline src/environment_windows.go]: pls rename to `err` for consistency, you can reuse it.

File: src/environment_windows.go [hunk 6]
```diff
@@ -105,31 +105,57 @@ func (env *environment) getCachePath() string {
...
+	lastSlash := strings.LastIndex(regPathParts[1], "\\")
+
+	regKey := regPathParts[1][lastSlash+1:]
+	regPath := regPathParts[1][0:lastSlash]
+
+	regKeyLogged := regKey
+	if regKeyLogged == "" {
+		regKeyLogged = "(Default)"
+	}
+	env.log(Error, "getWindowsRegistryKeyValue", fmt.Sprintf("getWindowsRegistryKeyValue: root:\"%s\", path:\"%s\", key:\"%s\"", regPathParts[0], regPath, regKeyLogged))
```
Comment (JanDeDobbeleer) [inline src/environment_windows.go]: why is this line here? Seems to always log that.

File: src/segment_winreg.go [hunk 7]
```diff
@@ -62,3 +74,45 @@ func (wr *winreg) templateString(segmentTemplate string) string {
 	}
 	return text
 }
+
+func (wr winreg) GetRegistryString(path string) (string, error) {
```
Comment (JanDeDobbeleer) [inline src/segment_winreg.go:78]: Oh boy do I want generics in go ASAP...

File: src/segment_winreg.go [hunk 8]
```diff
@@ -62,3 +74,45 @@ func (wr *winreg) templateString(segmentTemplate string) string {
...
+	regValue, err := wr.env.getWindowsRegistryKeyValue(path)
+
+	if regValue != nil {
```
Comment (JanDeDobbeleer) [inline src/segment_winreg.go]: Inverting this if allows the send if to be on the same level below.

**Review by JanDeDobbeleer — CHANGES_REQUESTED** [review]
A few remarks, nothing major. Pretty cool stuff tbh.

File: src/environment.go (hunk 1 above)
Comment (memcpy-rand-rand-rand) [reply src/environment.go:61]: I was wondering about that too - if possible to have the same interface, but on windows the windowsRegistryValueType and interface etc is full featured where on other platforms it's an interface with no info... I'll have a try.

File: src/environment_windows.go (hunk 2 above)
Comment (memcpy-rand-rand-rand) [reply src/environment_windows.go:140]: yes - I'm making the assumption that the split works - will add some validation.

File: src/environment_windows.go (hunk 3 above)
Comment (memcpy-rand-rand-rand) [reply src/environment_windows.go:148]: yup - validation needed.

File: src/environment_windows.go (hunk 6 above)
Comment (memcpy-rand-rand-rand) [reply src/environment_windows.go]: yes - mostly for debug implementation help.  Likely end-users wouldn't need it I guess.

File: src/segment_winreg.go (hunk 8 above)
Comment (memcpy-rand-rand-rand) [reply src/segment_winreg.go]: ooh, good spot - yeah, makes it a lot cleaner 👍

File: src/environment_windows.go (hunk 6 above)
Comment (memcpy-rand-rand-rand) [reply src/environment_windows.go]: ah whoops - I think I see what you mean, it's always logging an error not debug.

File: src/environment.go (hunk 1 above)
Comment (memcpy-rand-rand-rand) [reply src/environment.go:61]: ah - problem with that, segment_winreg.go has a dependency on that interface and enum now, so to make it windows-only at an environment level means the segment would then have to have different platform impls to avoid using the interface/enum.

Comment (JanDeDobbeleer) [reply src/environment.go:61]: Gotcha!

**Comment by memcpy-rand-rand-rand** [comment]
@JanDeDobbeleer I branched this and squashed commits to remove the commit lint issue; it has the changes as reviewed in this PR.  Would you prefer I create a PR on the new branch?

**Comment by JanDeDobbeleer** [comment]
@memcpy-rand-rand-rand works for me.
